api/src/main/java/org/pmiops/workbench/api/UserMetricsController.java

```diff
@@ -0,0 +1,118 @@
+import org.pmiops.workbench.db.dao.WorkspaceService;
+import org.pmiops.workbench.db.model.User;
+import org.pmiops.workbench.db.model.UserRecentResource;
+import org.pmiops.workbench.db.model.Workspace;
+import org.pmiops.workbench.firecloud.FireCloudService;
+import org.pmiops.workbench.model.Cohort;
+import org.pmiops.workbench.model.FileDetail;
+import org.pmiops.workbench.model.RecentResource;
+import org.pmiops.workbench.model.RecentResourceResponse;
+import org.pmiops.workbench.firecloud.model.WorkspaceResponse;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.http.ResponseEntity;
+import org.springframework.web.bind.annotation.RestController;
+
+import javax.inject.Provider;
+import java.util.List;
+import java.util.Map;
+import java.util.function.Function;
+import java.util.stream.Collectors;
+
+@RestController
+public class UserMetricsController implements UserMetricsApiDelegate {
+  Provider<User> userProvider;
+  UserRecentResourceService userRecentResourceService;
+  WorkspaceService workspaceService;
+  FireCloudService fireCloudService;
+  private static final String NOTEBOOKS_WORKSPACE_DIRECTORY = "notebooks/";
+  private final int DISTINCT_WORKSPACE_LIMIT = 5;
+
+  @Autowired
+  UserMetricsController(Provider<User> userProvider,
+      UserRecentResourceService userRecentResourceService,
+      WorkspaceService workspaceService,
+      FireCloudService fireCloudService) {
+    this.userProvider = userProvider;
+    this.userRecentResourceService = userRecentResourceService;
+    this.workspaceService = workspaceService;
+    this.fireCloudService = fireCloudService;
+  }
+
+  /**
+   * Gets the list of all resources recently access by user in order of access date time
+   *
+   * @return
+   */
+  @Override
+  public ResponseEntity<RecentResourceResponse> getUserRecentResources() {
+    long userId = userProvider.get().getUserId();
+    RecentResourceResponse recentResponse = new RecentResourceResponse();
+    List<UserRecentResource> userRecentResourceList = userRecentResourceService.findAllResourcesByUser(userId);
+    List<Long> workspaceIdList = userRecentResourceList
+        .stream()
+        .map(UserRecentResource::getWorkspaceId)
+        .distinct()
+        .limit(DISTINCT_WORKSPACE_LIMIT)
+        .collect(Collectors.toList());
+
+    Map<Long, String> workspaceAccessMap = workspaceIdList.stream().collect(Collectors.toMap(id -> id, id -> {
+      Workspace workspace = workspaceService.findByWorkspaceId((long) id);
+      WorkspaceResponse workspaceResponse = fireCloudService
+          .getWorkspace(workspace.getWorkspaceNamespace(),
+              workspace.getFirecloudName());
+      return workspaceResponse.getAccessLevel();
+    }));
+
+    userRecentResourceList.stream().forEach(userRecentResource -> {
+      RecentResource resource = TO_CLIENT.apply(userRecentResource);
+      resource.setPermission(workspaceAccessMap.get(userRecentResource.getWorkspaceId()));
```

Comment:
This needs to handle the fact that this workspace might not be present in the map (if so, you want to skip it)

Reply:
Ahhh sorry to miss this.

Reply:
@rushtong @jscherdin will it be better to use
 workspaceIdList.stream().parallel().collect(<permission>)

than restricting to 5 distinct workspace permission?